File: frontend/src/services/partnershipService.js
// frontend/src/services/partnershipService.js
import apiClient from '../utils/apiClient';

const API_BASE = '/partnerships';

/**
 * Récupère la liste des partenariats de l'entreprise connectée
 */
export const fetchPartnerships = async () => {
  try {
    const { data } = await apiClient.get(API_BASE);
    // La réponse est {data: [...]}, donc on accède à data.data
    const partnershipsList = data?.data || data || [];
    return Array.isArray(partnershipsList) ? partnershipsList : [];
  } catch (error) {
    console.error('Erreur lors de la récupération des partenariats:', error);
    throw error;
  }
};

/**
 * Récupère la liste des partenariats disponibles pour le transfert de courses
 * (uniquement ceux où l'entreprise est propriétaire)
 */
export const fetchPartnershipsForTransfer = async () => {
  try {
    const { data } = await apiClient.get(`${API_BASE}/for-transfer`);
    // La réponse est {data: [...]}, donc on accède à data.data
    const partnershipsList = data?.data || data || [];
    return Array.isArray(partnershipsList) ? partnershipsList : [];
  } catch (error) {
    console.error('Erreur lors de la récupération des partenariats pour transfert:', error);
    throw error;
  }
};

/**
 * Récupère un partenariat par son ID
 */
export const fetchPartnership = async (partnershipId) => {
  try {
    const { data } = await apiClient.get(`${API_BASE}/${partnershipId}`);
    return data;
  } catch (error) {
    console.error('Erreur lors de la récupération du partenariat:', error);
    throw error;
  }
};

/**
 * Crée un nouveau partenariat
 */
export const createPartnership = async (partnershipData) => {
  try {
    const { data } = await apiClient.post(API_BASE, partnershipData);
    return data;
  } catch (error) {
    console.error('Erreur lors de la création du partenariat:', error);
    throw error;
  }
};

/**
 * Met à jour un partenariat
 */
export const updatePartnership = async (partnershipId, partnershipData) => {
  try {
    const { data } = await apiClient.put(`${API_BASE}/${partnershipId}`, partnershipData);
    return data;
  } catch (error) {
    console.error('Erreur lors de la mise à jour du partenariat:', error);
    throw error;
  }
};

/**
 * Désactive un partenariat
 */
export const deactivatePartnership = async (partnershipId) => {
  try {
    const { data } = await apiClient.delete(`${API_BASE}/${partnershipId}`);
    return data;
  } catch (error) {
    console.error('Erreur lors de la désactivation du partenariat:', error);
    throw error;
  }
};

/**
 * Récupère les transferts d'un partenariat
 */
export const fetchPartnershipTransfers = async (partnershipId) => {
  try {
    const { data } = await apiClient.get(`${API_BASE}/${partnershipId}/transfers`);
    return Array.isArray(data) ? data : [];
  } catch (error) {
    console.error('Erreur lors de la récupération des transferts:', error);
    throw error;
  }
};

/**
 * Propose un transfert de course à un partenaire
 */
export const proposeTransfer = async (partnershipId, bookingId, transferModel = null) => {
  try {
    const payload = { booking_id: bookingId };
    if (transferModel) {
      payload.transfer_model = transferModel;
    }
    const { data } = await apiClient.post(`${API_BASE}/${partnershipId}/transfers`, payload);
    // La réponse est {data: {...}}, donc on accède à data.data
    return data?.data || data;
  } catch (error) {
    console.error('Erreur lors de la proposition du transfert:', error);
    throw error;
  }
};

/**
 * Accepte un transfert
 */
export const acceptTransfer = async (transferId) => {
  try {
    const { data } = await apiClient.post(`/partnerships/transfers/${transferId}/accept`);
    return data;
  } catch (error) {
    console.error('Erreur lors de l\'acceptation du transfert:', error);
    throw error;
  }
};

/**
 * Refuse un transfert
 */
export const rejectTransfer = async (transferId) => {
  try {
    const { data } = await apiClient.post(`/partnerships/transfers/${transferId}/reject`);
    return data;
  } catch (error) {
    console.error('Erreur lors du refus du transfert:', error);
    throw error;
  }
};

/**
 * Valide la complétion d'un transfert
 */
export const validateTransfer = async (transferId) => {
  try {
    const { data } = await apiClient.post(`/partnerships/transfers/${transferId}/validate`);
    return data;
  } catch (error) {
    console.error('Erreur lors de la validation du transfert:', error);
    throw error;
  }
};

/**
 * Récupère tous les transferts de l'entreprise connectée
 * Note: Pour l'instant, on récupère les transferts via les partenariats
 * Une route dédiée pourrait être ajoutée plus tard
 */
export const fetchAllTransfers = async (status = null) => {
  try {
    // Récupérer tous les partenariats
    const partnerships = await fetchPartnerships();
    
    // Récupérer les transferts de chaque partenariat
    const allTransfers = [];
    for (const partnership of partnerships) {
      try {
        const transfers = await fetchPartnershipTransfers(partnership.id);
        allTransfers.push(...transfers);
      } catch (err) {
        console.warn(`Erreur lors de la récupération des transferts pour le partenariat ${partnership.id}:`, err);
      }
    }
    
    // Filtrer par statut si demandé
    if (status) {
      return allTransfers.filter(t => t.status === status);
    }
    
    return allTransfers;
  } catch (error) {
    console.error('Erreur lors de la récupération des transferts:', error);
    throw error;
  }
};

/**
 * Récupère les statistiques d'un partenariat
 */
export const fetchPartnershipStats = async (partnershipId) => {
  try {
    const { data } = await apiClient.get(`${API_BASE}/${partnershipId}/stats`);
    return data;
  } catch (error) {
    console.error('Erreur lors de la récupération des statistiques:', error);
    throw error;
  }
};

/**
 * Récupère les factures mensuelles d'un partenariat
 */
export const fetchPartnershipInvoices = async (partnershipId) => {
  try {
    const { data } = await apiClient.get(`${API_BASE}/${partnershipId}/invoices`);
    return Array.isArray(data) ? data : [];
  } catch (error) {
    console.error('Erreur lors de la récupération des factures:', error);
    throw error;
  }
};

/**
 * Génère une facture mensuelle consolidée pour un partenariat
 */
export const generateMonthlyInvoice = async (partnershipId, year, month) => {
  try {
    const { data } = await apiClient.post(`${API_BASE}/${partnershipId}/invoices`, {
      year,
      month,
    });
    return data;
  } catch (error) {
    console.error('Erreur lors de la génération de la facture:', error);
    throw error;
  }
};

/**
 * Récupère les détails d'une facture partenaire
 */
export const fetchPartnerInvoice = async (partnershipId, invoiceId) => {
  try {
    const { data } = await apiClient.get(`${API_BASE}/${partnershipId}/invoices/${invoiceId}`);
    return data;
  } catch (error) {
    console.error('Erreur lors de la récupération de la facture:', error);
    throw error;
  }
};

/**
 * Marque une facture partenaire comme payée
 */
export const markInvoiceAsPaid = async (partnershipId, invoiceId) => {
  try {
    const { data } = await apiClient.put(`${API_BASE}/${partnershipId}/invoices/${invoiceId}`);
    return data;
  } catch (error) {
    console.error('Erreur lors de la mise à jour de la facture:', error);
    throw error;
  }
};

/**
 * Recherche des entreprises par nom
 */
export const searchCompanies = async (query) => {
  try {
    const { data } = await apiClient.get(`${API_BASE}/search-companies`, {
      params: { q: query },
    });
    return Array.isArray(data) ? data : [];
  } catch (error) {
    console.error('Erreur lors de la recherche d\'entreprises:', error);
    throw error;
  }
};

/**
 * Récupère les demandes de partenariat en attente
 */
export const fetchPendingRequests = async () => {
  try {
    const { data } = await apiClient.get(`${API_BASE}/requests`);
    return Array.isArray(data) ? data : [];
  } catch (error) {
    console.error('Erreur lors de la récupération des demandes:', error);
    throw error;
  }
};

/**
 * Accepter une demande de partenariat
 */
export const acceptPartnershipRequest = async (partnershipId) => {
  try {
    const { data } = await apiClient.post(`${API_BASE}/${partnershipId}/accept`);
    return data;
  } catch (error) {
    console.error('Erreur lors de l\'acceptation de la demande:', error);
    throw error;
  }
};

/**
 * Refuser une demande de partenariat
 */
export const rejectPartnershipRequest = async (partnershipId) => {
  try {
    const { data } = await apiClient.post(`${API_BASE}/${partnershipId}/reject`);
    return data;
  } catch (error) {
    console.error('Erreur lors du refus de la demande:', error);
    throw error;
  }
};

File: backend/services/partnership_service.py
# services/partnership_service.py
"""Service pour gérer les partenariats entre entreprises."""

from ext import db
from models.company import Company
from models.enums import PartnershipStatus, TransferModel
from models.partnership import Partnership


class PartnershipService:
    """Service pour la gestion des partenariats."""

    @staticmethod
    def create_partnership(
        owner_company_id: int,
        partner_company_id: int,
        default_transfer_model: TransferModel = TransferModel.SUBCONTRACT,
        default_margin_percent: float | None = None,
        default_partner_tariff_percent: float | None = None,
        auto_accept: bool = False,
        auto_invoice: bool = True,
        payment_terms_days: int = 30,
    ) -> Partnership:
        """Créer un nouveau partenariat entre deux entreprises.

        Args:
            owner_company_id: ID de l'entreprise propriétaire
            partner_company_id: ID de l'entreprise partenaire
            default_transfer_model: Modèle de transfert par défaut
            default_margin_percent: Marge que l'entreprise propriétaire garde (ex: 20%)
            default_partner_tariff_percent: % du prix client pour le partenaire (ex: 80%)
            auto_accept: Auto-acceptation des transferts
            auto_invoice: Facturation automatique
            payment_terms_days: Délai de paiement en jours

        Returns:
            Partnership créé

        Raises:
            ValueError: Si les entreprises sont identiques ou si le partenariat existe déjà
        """
        if owner_company_id == partner_company_id:
            raise ValueError("Une entreprise ne peut pas être partenaire d'elle-même")

        # Vérifier si une demande existe déjà (pending ou accepted)
        existing = (
            Partnership.query.filter_by(
                owner_company_id=owner_company_id,
                partner_company_id=partner_company_id,
            )
            .filter(
                Partnership.status.in_(
                    [PartnershipStatus.PENDING, PartnershipStatus.ACCEPTED]
                )
            )
            .first()
        )

        if existing:
            if existing.status == PartnershipStatus.PENDING:
                raise ValueError("Une demande de partenariat est déjà en attente")
            raise ValueError("Ce partenariat existe déjà")

        # Vérifier que les entreprises existent
        owner_company = Company.query.get(owner_company_id)
        if not owner_company:
            raise ValueError(f"Entreprise propriétaire {owner_company_id} introuvable")

        partner_company = Company.query.get(partner_company_id)
        if not partner_company:
            raise ValueError(f"Entreprise partenaire {partner_company_id} introuvable")

        # SQLAlchemy 2.0 avec Mapped nécessite d'assigner les attributs après création
        partnership = Partnership()
        partnership.owner_company_id = owner_company_id
        partnership.partner_company_id = partner_company_id
        partnership.default_transfer_model = default_transfer_model
        from decimal import Decimal

        partnership.default_margin_percent = (
            Decimal(str(default_margin_percent))
            if default_margin_percent is not None
            else None
        )
        partnership.default_partner_tariff_percent = (
            Decimal(str(default_partner_tariff_percent))
            if default_partner_tariff_percent is not None
            else None
        )
        partnership.auto_accept_rules = auto_accept
        partnership.auto_invoice = auto_invoice
        partnership.payment_terms_days = payment_terms_days
        partnership.status = PartnershipStatus.PENDING  # Demande en attente

        db.session.add(partnership)
        db.session.commit()
        return partnership

    @staticmethod
    def get_partnerships_for_company(company_id: int) -> list[Partnership]:
        """Récupérer tous les partenariats d'une entreprise (en tant que propriétaire ou partenaire).

        Args:
            company_id: ID de l'entreprise

        Returns:
            Liste des partenariats actifs
        """
        return (
            Partnership.query.filter(
                (Partnership.owner_company_id == company_id)
                | (Partnership.partner_company_id == company_id)
            )
            .filter_by(status=PartnershipStatus.ACCEPTED, is_active=True)
            .all()
        )

    @staticmethod
    def get_owner_partnerships_for_company(company_id: int) -> list[Partnership]:
        """Récupérer les partenariats où l'entreprise est propriétaire (pour transfert de courses).

        Args:
            company_id: ID de l'entreprise

        Returns:
            Liste des partenariats actifs où l'entreprise est propriétaire
        """
        return Partnership.query.filter_by(
            owner_company_id=company_id,
            status=PartnershipStatus.ACCEPTED,
            is_active=True,
        ).all()

    @staticmethod
    def get_partnership(
        owner_company_id: int, partner_company_id: int
    ) -> Partnership | None:
        """Récupérer un partenariat spécifique.

        Args:
            owner_company_id: ID de l'entreprise propriétaire
            partner_company_id: ID de l'entreprise partenaire

        Returns:
            Partnership ou None si non trouvé
        """
        return Partnership.query.filter_by(
            owner_company_id=owner_company_id,
            partner_company_id=partner_company_id,
            is_active=True,
        ).first()

    @staticmethod
    def get_partnership_by_id(partnership_id: int) -> Partnership | None:
        """Récupérer un partenariat par son ID.

        Args:
            partnership_id: ID du partenariat

        Returns:
            Partnership ou None si non trouvé
        """
        return Partnership.query.get(partnership_id)

    @staticmethod
    def update_partnership(
        partnership_id: int,
        default_transfer_model: TransferModel | None = None,
        default_margin_percent: float | None = None,
        default_partner_tariff_percent: float | None = None,
        auto_accept: bool | None = None,
        auto_invoice: bool | None = None,
        payment_terms_days: int | None = None,
        status: PartnershipStatus | None = None,
    ) -> Partnership:
        """Mettre à jour un partenariat.

        Args:
            partnership_id: ID du partenariat
            default_transfer_model: Nouveau modèle de transfert
            default_margin_percent: Nouvelle marge
            default_partner_tariff_percent: Nouveau tarif partenaire
            auto_accept: Nouvelle règle d'auto-acceptation
            auto_invoice: Nouvelle règle de facturation automatique
            payment_terms_days: Nouveau délai de paiement

        Returns:
            Partnership mis à jour

        Raises:
            ValueError: Si le partenariat n'existe pas
        """
        partnership = Partnership.query.get(partnership_id)
        if not partnership:
            raise ValueError(f"Partenariat {partnership_id} introuvable")

        if default_transfer_model is not None:
            partnership.default_transfer_model = default_transfer_model
        if default_margin_percent is not None:
            partnership.default_margin_percent = default_margin_percent
        if default_partner_tariff_percent is not None:
            partnership.default_partner_tariff_percent = default_partner_tariff_percent
        if auto_accept is not None:
            partnership.auto_accept_rules = auto_accept
        if auto_invoice is not None:
            partnership.auto_invoice = auto_invoice
        if payment_terms_days is not None:
            partnership.payment_terms_days = payment_terms_days
        if status is not None:
            partnership.status = status

        db.session.commit()
        return partnership

    @staticmethod
    def deactivate_partnership(partnership_id: int) -> Partnership:
        """Désactiver un partenariat.

        Args:
            partnership_id: ID du partenariat

        Returns:
            Partnership désactivé

        Raises:
            ValueError: Si le partenariat n'existe pas
        """
        partnership = Partnership.query.get(partnership_id)
        if not partnership:
            raise ValueError(f"Partenariat {partnership_id} introuvable")

        partnership.is_active = False
        db.session.commit()
        return partnership

    @staticmethod
    def delete_partnership(partnership_id: int, company_id: int) -> None:
        """Supprime complètement un partenariat.

        Args:
            partnership_id: ID du partenariat
            company_id: ID de l'entreprise qui demande la suppression

        Raises:
            ValueError: Si le partenariat n'existe pas ou si l'entreprise n'est pas autorisée
        """
        import logging

        logger = logging.getLogger(__name__)

        partnership = Partnership.query.get(partnership_id)
        if not partnership:
            logger.warning(
                "Tentative de suppression d'un partenariat inexistant: %s",
                partnership_id,
            )
            raise ValueError(f"Partenariat {partnership_id} introuvable")

        # Vérifier que l'entreprise est liée au partenariat
        if company_id not in {
            partnership.owner_company_id,
            partnership.partner_company_id,
        }:
            logger.warning(
                "Tentative de suppression non autorisée: company_id=%s, partnership_id=%s",
                company_id,
                partnership_id,
            )
            raise ValueError("Vous n'êtes pas autorisé à supprimer ce partenariat")

        try:
            # Vérifier s'il y a des transferts ou factures liés
            from models.booking_transfer import BookingTransfer
            from models.partner_invoice import PartnerInvoice

            transfers_count = BookingTransfer.query.filter_by(
                partnership_id=partnership_id
            ).count()
            invoices_count = PartnerInvoice.query.filter_by(
                partnership_id=partnership_id
            ).count()

            if transfers_count > 0 or invoices_count > 0:
                logger.warning(
                    (
                        "Tentative de suppression d'un partenariat avec des dépendances: "
                        "partnership_id=%s, transfers=%s, invoices=%s"
                    ),
                    partnership_id,
                    transfers_count,
                    invoices_count,
                )
                # Note: Avec ondelete="CASCADE", ces enregistrements seront supprimés automatiquement
                # mais on log quand même pour information

            # Supprimer le partenariat de la base de données
            db.session.delete(partnership)
            db.session.commit()
            logger.info(
                "Partenariat %s supprimé avec succès par company_id=%s",
                partnership_id,
                company_id,
            )
        except Exception as e:
            db.session.rollback()
            logger.exception(
                "Erreur lors de la suppression du partenariat %s: %s", partnership_id, e
            )
            raise

    @staticmethod
    def accept_partnership_request(partnership_id: int, company_id: int) -> Partnership:
        """Accepter une demande de partenariat.

        Args:
            partnership_id: ID du partenariat
            company_id: ID de l'entreprise qui accepte (doit être le partenaire)

        Returns:
            Partnership accepté

        Raises:
            ValueError: Si le partenariat n'existe pas ou si l'entreprise n'est pas autorisée
        """
        partnership = Partnership.query.get(partnership_id)
        if not partnership:
            raise ValueError(f"Partenariat {partnership_id} introuvable")

        if partnership.partner_company_id != company_id:
            raise ValueError(
                "Seule l'entreprise partenaire peut accepter cette demande"
            )

        if partnership.status != PartnershipStatus.PENDING:
            raise ValueError("Cette demande ne peut plus être acceptée")

        partnership.status = PartnershipStatus.ACCEPTED
        partnership.is_active = True
        db.session.commit()
        return partnership

    @staticmethod
    def reject_partnership_request(partnership_id: int, company_id: int) -> Partnership:
        """Refuser une demande de partenariat.

        Args:
            partnership_id: ID du partenariat
            company_id: ID de l'entreprise qui refuse (doit être le partenaire)

        Returns:
            Partnership refusé

        Raises:
            ValueError: Si le partenariat n'existe pas ou si l'entreprise n'est pas autorisée
        """
        partnership = Partnership.query.get(partnership_id)
        if not partnership:
            raise ValueError(f"Partenariat {partnership_id} introuvable")

        if partnership.partner_company_id != company_id:
            raise ValueError("Seule l'entreprise partenaire peut refuser cette demande")

        if partnership.status != PartnershipStatus.PENDING:
            raise ValueError("Cette demande ne peut plus être refusée")

        partnership.status = PartnershipStatus.REJECTED
        partnership.is_active = False
        db.session.commit()
        return partnership

    @staticmethod
    def get_pending_requests_for_company(company_id: int) -> list[Partnership]:
        """Récupérer les demandes de partenariat en attente pour une entreprise.

        Args:
            company_id: ID de l'entreprise (en tant que partenaire)

        Returns:
            Liste des demandes en attente
        """
        return Partnership.query.filter_by(
            partner_company_id=company_id,
            status=PartnershipStatus.PENDING,
        ).all()
